File: java/graph/warshall.java
/*Floyd-Warshall Algorithm is an algorithm for finding the shortest path between all the pairs of vertices in a weighted graph.
This algorithm works for both the directed and undirected weighted graphs. If there is no direct distance between two vertices
then it is considered as -1*/

import java.util.*;

public class warshall
{
    int size; // the size of the graph
    
    void warshall_algorithm(int[][] g)
    {
        int i, j, k;
        int[][] dist=new int[size][size]; // A 2D array to store distance
        for(i=0;i<size;i++)
        {
            for(j=0;j<size;j++)
            {
                dist[i][j]=g[i][j];
            }
        }
        for(k=0;k<size;k++)
        {
            for(i=0;i<size;i++)
            {
                for(j=0;j<size;j++)
                {
                    if((dist[i][k]+dist[k][j]<dist[i][j]) && (dist[i][k]!=-1) && (dist[k][j]!=-1)) // Finding the smallest distance between two vertexes if it exists
                        dist[i][j]=dist[i][k]+dist[k][j];
                }
            }
        }
        System.out.println("The distance graph is: \n"); // Printing the 2D array storing distance
        for(i=0;i<size;i++)
        {
            for(j=0;j<size;j++)
            {
                if(dist[i][j]>=0)
                System.out.print(dist[i][j]+" ");
                else
                System.out.print("-1 ");
            }
            System.out.println();
        }
    }

    public static void main(String args[])
    {
        Scanner sc=new Scanner(System.in);
        warshall obj=new warshall(); //Creation of an object
        System.out.println("Enter the number of vertices: ");
        obj.size=sc.nextInt();
        int[][] graph=new int[obj.size][obj.size];
        System.out.println("Enter the graph: ");
        for(int i=0;i<obj.size;i++)
        {
            for(int j=0;j<obj.size;j++)
            {
                graph[i][j]=sc.nextInt();
            }
        }
        obj.warshall_algorithm(graph); // Calling the Warshall algorithm function
    }
}

/*
Time Complexity:O(size^3)
Space Complexity:O(size^2)
Input/Output:-
Enter the number of vertices: 
4
Enter the graph: 
1
2
3
4
1
5
3
-3
-2
0
-1
5
-4
-3
2
-1
The distance graph is: 

1 2 3 -1 
-1 -1 3 -1 
-1 -1 -1 -1 
-1 -1 -1 -1 
*/

File: c++/graph/warshall.cpp
#include <iostream>
#include <climits>
#include <iomanip>
using namespace std;
// defined number of vertices
#define N 4
#define M INT_MAX
void printpath(int path[][N], int v, int u)
{
    if (path[v][u] == v)
        return;

    printpath(path, v, path[v][u]);
    cout << path[v][u] << " ";
}
void solution(int cost[N][N], int path[N][N])
{
    for (int v = 0; v < N; v++)
    {
        for (int u = 0; u < N; u++)
        {
            if (u != v && path[v][u] != -1)
            {
                cout << "Shortest Path from " << v << " -> " << u << " is ("
                     << v << " ";
                printpath(path, v, u);
                cout << u << ")" << endl;
            }
        }
    }
}
// implemented Floyd warshall Algorithm
void WarshallAlgo(int adjMatrix[][N])
{
    int cost[N][N], path[N][N];
    for (int v = 0; v < N; v++)
    {
        for (int u = 0; u < N; u++)
        {
            cost[v][u] = adjMatrix[v][u];

            if (v == u)
                path[v][u] = 0;
            else if (cost[v][u] != INT_MAX)
                path[v][u] = v;
            else
                path[v][u] = -1;
        }
    }
    // Adding vertices individually
    for (int k = 0; k < N; k++)
    {
        for (int v = 0; v < N; v++)
        {
            for (int u = 0; u < N; u++)
            {
                if (cost[v][k] != INT_MAX && cost[k][u] != INT_MAX && cost[v][k] + cost[k][u] < cost[v][u])
                {
                    cost[v][u] = cost[v][k] + cost[k][u];
                    path[v][u] = path[k][u];
                }
            }

            if (cost[v][v] < 0)
            {
                cout << "Negative Weight Cycle Found!!";
                return;
            }
        }
    }
    solution(cost, path);
}

int main()
{
    int adjMatrix[N][N] =
        {
            {0, M, -2, M},
            {4, 0, 3, M},
            {M, M, 0, 2},
            {M, -1, M, 0}};  // graph input
    WarshallAlgo(adjMatrix); // calling of function

    return 0;
}
/*
Time Complexity:O(Ver^3)
Space Complexity:O(Ver^2)
Shortest Path from 0 -> 1 is (0 2 3 1)
Shortest Path from 0 -> 2 is (0 2)
Shortest Path from 0 -> 3 is (0 2 3)
Shortest Path from 1 -> 0 is (1 0)
Shortest Path from 1 -> 2 is (1 0 2)
Shortest Path from 1 -> 3 is (1 0 2 3)
Shortest Path from 2 -> 0 is (2 3 1 0)
Shortest Path from 2 -> 1 is (2 3 1)
Shortest Path from 2 -> 3 is (2 3)
Shortest Path from 3 -> 0 is (3 1 0)
Shortest Path from 3 -> 1 is (3 1)
Shortest Path from 3 -> 2 is (3 1 0 2)*/